asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py

```diff
@@ -0,0 +1,926 @@
+    """
+    Pre-processes raw historical installers data by:
+    - Dropping duplicate instances (if they exist);
+    - Renaming columns;
+    - Joining installation and design variables;
+    - Adding instances for installers with installations but with info missing from installers table;
+    - Creating flag variables for whether an installer is certified for a certain technology;
+    - Geocoding data by mapping postcode to latitude and longitude values;
+    - Creating a variable that uniquely identifies installers;
+    - Adding certification body information.
+
+    Args:
+        raw_historical_installers: raw historical installers data
+        raw_historical_installations: raw historical installations data
+        geographical_data: geographical data with postcode, latitude and longitude information
+        companies_house_api_key: Companies House API key
+    Returns:
+        Processed historical installers data
+    """
+
+    # Apply basic preprocessing to table of installations
+    basic_preprocessing_of_installations(raw_historical_installations)
+
+    # Dropping duplicate lines
+    raw_historical_installers.drop_duplicates(inplace=True)
+
+    # Renaming columns
+    raw_historical_installers.columns = rename_columns(
+        raw_historical_installers.columns
+    )
+
+    # Joining installation and design variables
+    join_installation_and_design_vars(raw_historical_installers)
+
+    # flag "original_records" (if originally in installers or inputed from installations)
+    raw_historical_installers["original_record"] = [
+        True for i in range(len(raw_historical_installers))
+    ]
+
+    # Get info about installers with installations but with info missing from installers
+    missing_installers = get_missing_installers_info(
+        raw_historical_installations, raw_historical_installers, companies_house_api_key
+    )
+
+    raw_historical_installers = pd.concat(
+        [raw_historical_installers, missing_installers]
+    )
+    raw_historical_installers.reset_index(drop=True, inplace=True)
+
+    # Updating full_address variable
+    raw_historical_installers["full_address"] = raw_historical_installers.apply(
+        lambda x: update_full_address(
+            x["address_1"],
+            x["address_2"],
+            x["town"],
+            x["county"],
+            x["postcode"],
+            x["original_record"],
+            x["full_address"],
+        ),
+        axis=1,
+    )
+
+    # Creating flag for whether certified for a certain technology
+    create_certified_flags(raw_historical_installers)
+
+    # Geocoding data
+    raw_historical_installers = geocode_postcode(
+        raw_historical_installers, geographical_data
+    )
+
+    # Create installer unique ID
+    create_installer_unique_id(raw_historical_installers)
+
+    # Certification body information
+    raw_historical_installers = add_certification_body_info(
+        raw_historical_installers, raw_historical_installations
```

Comment:
Line 793 adds 'certification_body' to installations dataframe. Lines 812 to 818 adds 'certification_body' to installers dataframe. Resulting in duplicate 'certification_body' in merge. Please see suggestion referred in this function's definition (Lines 752 to 759) to account for this.

Reply:
see my comment above :)